File: ios/Utils/Factory/TypedBannerAdSizeFactory.swift
import Foundation
import YandexMobileAds

class TypedBannerAdSizeFactory {
    func createSticky(width: CGFloat) -> TypedBannerAdSize {
        let bannerAdSize = BannerAdSize.stickySize(withContainerWidth: width)
        return TypedBannerAdSize(
            bannerAdSize: bannerAdSize,
            initialWidth: width,
            initialHeight: bannerAdSize.size.height,
            type: Constants.BannerType.sticky
        )
    }
    
    func createInline(width: CGFloat, maxHeight: CGFloat) -> TypedBannerAdSize {
        let bannerAdSize = BannerAdSize.inlineSize(withWidth: width, maxHeight: maxHeight)
        return TypedBannerAdSize(
            bannerAdSize: bannerAdSize,
            initialWidth: width,
            initialHeight: maxHeight,
            type: Constants.BannerType.inline
        )
    }
    
    func createFromDictionary(_ dictionary: [String: Any]) -> TypedBannerAdSize? {
        guard let type = dictionary[Constants.Key.bannerType] as? String else {
            return nil
        }
        switch type {
        case Constants.BannerType.sticky:
            guard let width = dictionary[Constants.Key.initialWidth] as? CGFloat else {
                return nil
            }
            return createSticky(width: width)
        case Constants.BannerType.inline:
            guard let width = dictionary[Constants.Key.initialWidth] as? CGFloat,
                  let height = dictionary[Constants.Key.initialHeight] as? CGFloat
            else {
                return nil
            }
            return createInline(width: width, maxHeight: height)
        default:
            return nil
        }
    }
}

File: android/src/main/java/com/yandexmobileads/util/factory/TypedBannerAdSizeFactory.kt
package com.yandexmobileads.util.factory

import android.content.Context
import com.facebook.react.bridge.ReadableMap
import com.yandex.mobile.ads.banner.BannerAdSize
import com.yandexmobileads.banner.TypedBannerAdSize
import com.yandexmobileads.common.Constants

internal class TypedBannerAdSizeFactory(
    private val context: Context,
) {

    fun createSticky(width: Int): TypedBannerAdSize {
        val bannerAdSize = BannerAdSize.stickySize(context, width)
        return TypedBannerAdSize(bannerAdSize, Constants.BannerType.STICKY)
    }

    fun createInline(width: Int, maxHeight: Int): TypedBannerAdSize {
        val bannerAdSize = BannerAdSize.inlineSize(context, width, maxHeight)
        return TypedBannerAdSize(bannerAdSize, Constants.BannerType.INLINE)
    }

    fun createFromReadableMap(readableMap: ReadableMap): TypedBannerAdSize? {
        val type = readableMap.getString(Constants.Key.BANNER_TYPE)
        return when (type) {
            Constants.BannerType.STICKY -> {
                if (readableMap.hasKey(Constants.Key.INITIAL_WIDTH).not()) {
                    return null
                }
                val width = readableMap.getInt(Constants.Key.INITIAL_WIDTH)
                createSticky(width)
            }

            Constants.BannerType.INLINE -> {
                if (readableMap.hasKey(Constants.Key.INITIAL_WIDTH).not()) {
                    return null
                }
                if (readableMap.hasKey(Constants.Key.INITIAL_HEIGHT).not()) {
                    return null
                }
                val width = readableMap.getInt(Constants.Key.INITIAL_WIDTH)
                val maxHeight = readableMap.getInt(Constants.Key.INITIAL_HEIGHT)
                createInline(width, maxHeight)
            }

            else -> null
        }
    }
}
